flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/Executing.java

```diff
@@ -0,0 +1,261 @@
+import org.apache.flink.runtime.executiongraph.TaskExecutionStateTransition;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+
+/** State which represents a running job with an {@link ExecutionGraph} and assigned slots. */
+class Executing extends StateWithExecutionGraph implements ResourceConsumer {
+
+    private final Context context;
+
+    private final ClassLoader userCodeClassLoader;
+
+    Executing(
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger,
+            Context context,
+            ClassLoader userCodeClassLoader) {
+        super(context, executionGraph, executionGraphHandler, operatorCoordinatorHandler, logger);
+        this.context = context;
+        this.userCodeClassLoader = userCodeClassLoader;
+    }
+
+    @Override
+    public void onEnter() {
+        deploy();
+    }
+
+    @Override
+    public void cancel() {
+        context.goToCanceling(
+                getExecutionGraph(), getExecutionGraphHandler(), getOperatorCoordinatorHandler());
+    }
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        handleAnyFailure(cause);
+    }
+
+    private void handleAnyFailure(Throwable cause) {
+        final FailureResult failureResult = context.howToHandleFailure(cause);
+
+        if (failureResult.canRestart()) {
+            context.goToRestarting(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getBackoffTime());
+        } else {
+            context.goToFailing(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getFailureCause());
+        }
+    }
+
+    @Override
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionState) {
+        final boolean successfulUpdate = getExecutionGraph().updateState(taskExecutionState);
+
+        if (successfulUpdate) {
+            if (taskExecutionState.getExecutionState() == ExecutionState.FAILED) {
+                Throwable cause = taskExecutionState.getError(userCodeClassLoader);
+                handleAnyFailure(cause);
+            }
+        }
+
+        return successfulUpdate;
+    }
+
+    @Override
+    void onTerminalState(JobStatus terminalState) {
+        context.goToFinished(ArchivedExecutionGraph.createFrom(getExecutionGraph()));
+    }
+
+    private void deploy() {
+        final ExecutionGraph executionGraph = getExecutionGraph();
+        executionGraph.transitionToRunning();
```

Comment:
Can't we assume that an `ExecutionGraph` is in state `RUNNING` when it is being passed to a `StateWithExecutionGraph`? This simplifies things a bit.